GitUI/CommandsDialogs/FileStatusListContextMenuController.cs

```diff
@@ -103,5 +115,53 @@ public bool ShouldDisplayMenuSelectedParentToLocal(ContextMenuDiffToolInfo selec
             // Not visible if same revision as ShouldShowMenuFirstToLocal()
             return !selectionInfo.FirstIsParent;
         }
+
+        public bool ShouldEnableFirstSpecialCompare(FileStatusItem item)
+        {
+            // First item must be a git reference, i.e. other than work tree
+            return item != null && !item.Item.IsSubmodule && item.SecondRevision.ObjectId != ObjectId.WorkTreeId;
+        }
+
+        public bool ShouldEnableSecondSpecialCompare(FileStatusItem item)
+        {
+            // Work tree file must exist on file system
+            return item != null && !item.Item.IsSubmodule && (item.SecondRevision.ObjectId != ObjectId.WorkTreeId || !item.Item.IsDeleted);
+        }
+
+        /// <inheritdoc/>>
+        public string GetGitCommit([CanBeNull] GitModule module, [CanBeNull] FileStatusItem item, bool isFirst)
```

Comment:
When `GitModule` is passed around it makes me very uneasy. This adds unnecessary coupling.
How about this?
```suggestion
        public string GetGitCommit([CanBeNull] FileStatusItem item, [CanBeNull] Func<string, ObjectId, ObjectId> getFileBlobHash, bool isFirst)
```

Call-sites should be:
```cs
var first = _revisionDiffContextMenuController.GetGitCommit(diffFiles[1], Module.GetFileBlobHash, true);
```

And this should simplify testing as well.

Reply:
will change